docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,388 @@
+
+1. **Failure Recovery**: In case of a failure, the OS should be able to revert to the previous version.
+2. **Retry Configuration**: Support for configuring the number of update retries.
+3. **Registry Configuration**: Ability to specify the registry from which to pull the OS image.
+
+### Update Strategies
+
+Gardener will introduce two additional update strategies, `InPlaceUpdate` and `InPlaceUpdateOnLabel`, for worker pools, in addition to the existing `RollingUpdate`. The two new strategies are described as follows:
+
+#### InplaceUpdate strategy
+
+1. On shoot spec change for Kubernetes/Machine image version, `machine-controller-manager` will label all the nodes with `candidate-for-update` label. This label can be used to identify nodes that will be selected for updates in the future.
+2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent it from being scaled down by [`cluster-autoscaler`](https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider/cluster-autoscaler) during the update process.
+3. `machine-controller-manager` will select the nodes based on the configured `maxUnavailable` and label them with `selected-for-update`. This label can be used to identify nodes that are currently undergoing an update.
+4. `machine-controller-manager` will cordon and drain the node and label it with `ready-for-update` label.
+5. `gardener-node-agent` will detect the `ready-for-update` label and perform the update on the node. And it will wait for the update to complete within the specified timeout.
+6. Once the Kubernetes/Machine image version is updated, the node will be labeled by the `gardener-node-agent` with `update-successful` label.
+7. `machine-controller-manager` will `uncordon` the node, and remove all the added labels. The node is then ready to schedule the workloads.
+
+![InplaceUpdate](assets/inplace-update.png "InPlace Update Strategy")
+
+#### InplaceUpdateOnLabel strategy
+
+1. On shoot spec change, nodes will be labeled with `candidate-for-update`.
+2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent it from being scaled down by `cluster-autoscaler` during the update process.
+3. `machine-controller-manager` will wait for the `selected-for-update` label on the node. The orchestration of the update is solely the responsibility of the user. They are free to select the nodes to be updated at will.
+4. Once the `selected-for-update` label is added by the user on the node, steps 4-7 of the `Inplace` strategy will performed.
+
+![InplaceUpdateOnLabel](assets/inplace-update-onlabel.png "InPlace Update OnLabel Strategy")
+
+### Gardener
+
+#### `CloudProfile` API
+
+A new field `minSupportedVersionInPlaceUpdate` is introduced under `.spec.machineImages[]` under the CloudProfile spec. This field can be used to specify the minimum supported version from which the current machine image version can be in-place updated to.
```

Comment:
Minor nit: not sure why `supported` adjective is necessary in the above. It is implied implicitly.  Why can't it simply be `minVersionForInPlaceUpdate`

Reply:
In any case, please explain whether "we" ("the system") will calculate "intermediate upgrades" or not and the user has to find out and "plan" the "intermediate upgrades".

That was the main point of this review comment, because it wasn't clear to me when reading the text. I therefore wrote:
> ...maybe you had something else in mind than (3), then write that, e.g. prevent updating to a new version if the constraint would be violated (admission plugin).